Source/WebCore/dom/InternalObserverFirst.cpp

```diff
@@ -0,0 +1,113 @@
+*    documentation and/or other materials provided with the distribution.
+*
+* THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+* AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+* THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+* PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+* BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+* CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+* SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+* INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+* CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+* ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+* THE POSSIBILITY OF SUCH DAMAGE.
+*/
+
+#include "config.h"
+#include "InternalObserverFirst.h"
+
+#include "AbortController.h"
+#include "AbortSignal.h"
+#include "Exception.h"
+#include "ExceptionCode.h"
+#include "InternalObserver.h"
+#include "JSDOMPromiseDeferred.h"
+#include "Observable.h"
+#include "ScriptExecutionContext.h"
+#include "SubscribeOptions.h"
+#include "Subscriber.h"
+#include "SubscriberCallback.h"
+#include <JavaScriptCore/JSCJSValueInlines.h>
+
+namespace WebCore {
+
+class InternalObserverFirst final : public InternalObserver {
+public:
+    static Ref<InternalObserverFirst> create(ScriptExecutionContext& context, Ref<AbortSignal> signal, Ref<DeferredPromise>&& promise)
+    {
+        Ref internalObserver = adoptRef(*new InternalObserverFirst(context, signal, WTFMove(promise)));
+        internalObserver->suspendIfNeeded();
+        return internalObserver;
+    }
+
+private:
+    void next(JSC::JSValue value) final
+    {
+        (Ref { m_promise })->resolve<IDLAny>(value);
+        (Ref { m_signal })->signalAbort(JSC::jsUndefined());
+    }
+
+    void error(JSC::JSValue value) final
+    {
+        (Ref { m_promise })->reject<IDLAny>(value);
+    }
+
+    void complete() final
+    {
+        InternalObserver::complete();
+        (Ref { m_promise })->reject(Exception { ExceptionCode::RangeError, "No values in Observable"_s });
+    }
+
+    void visitAdditionalChildren(JSC::AbstractSlotVisitor&) const final
+    {
+    }
+
+    void visitAdditionalChildren(JSC::SlotVisitor&) const final
+    {
+    }
+
+    InternalObserverFirst(ScriptExecutionContext& context, Ref<AbortSignal> signal, Ref<DeferredPromise>&& promise)
+        : InternalObserver(context)
+        , m_signal(signal)
+        , m_promise(WTFMove(promise))
+    {
+    }
+
+    Ref<AbortSignal> m_signal;
+    Ref<DeferredPromise> m_promise;
+};
+
+void createInternalObserverOperatorFirst(ScriptExecutionContext& context, Ref<Observable> observable, SubscribeOptions options, Ref<DeferredPromise>&& promise)
+{
+    Ref signal = AbortSignal::create(&context);
+    Vector<Ref<AbortSignal>> dependentSignals = { WTFMove(signal) };
```

Comment:
@darinadler I applied your suggestion for the inline signal here. Though left it as 2 lines, to reenforce my `FIXME` comment below.

Do not, that when I pass the signal as suggested below together with the WTFMove, I do get a runtime crash like:

```
ASSERTION FAILED: m_ptr
12:19:36.125 3412   WebKit/WebKitBuild/Debug/usr/local/include/wtf/Ref.h(135) : T &WTF::Ref<WebCore::AbortSignal>::get() const [T = WebCore::AbortSignal, _PtrTraits = WTF::RawPtrTraits<WebCore::AbortSignal>, RefDerefTraits = WTF::DefaultRefDerefTraits<WebCore::AbortSignal>]
```

Reply:
Sounds like you did something different than what I meant. What I meant was:
```
Vector<Ref<AbortSignal>> dependentSignals { AbortSignal::create(&context) };
```

Reply:
Yeah. That would work, given how [I am not passing the right signal](https://github.com/WebKit/WebKit/pull/35261#discussion_r1809770841). But then, whats the point in creating an inline signal in the first place if its not used at all.

But when passing the right signal down, and your suggestion — I do get that crash.

Reply:
The assertion failure you mention above is what you'd get if you use `signal` after moving from it. You can only move a value once and must not use it after moving. I don't know exactly where the error is but I'm sure we can find it.